exploration/0474-text-vs-code.md

```diff
@@ -0,0 +1,204 @@
+
+The current syntax and active proposals include some sigil + name combinations,
+such as `:number`, `$var`, `|literal|`, `+bold`, and `@attr`.
+
+The current syntax supports unquoted literal values as operands.
+
+## Proposed Design
+
+TBD
+
+## Alternatives Considered
+
+### Start in code, encapsulate text
+
+This approach treats messages as something like a resource format for pattern values.
+Keywords are declared directly at the top level of a message,
+and patterns are surrounded by `{...}`.
+
+Whitespace in patterns is never trimmed.
+
+Some code statements (variable declarations and match statements)
+also use `{...}` to surround values at the top level,
+so counting `{` instances is not sufficient to identify if a value is "code" or "text".
+
+The `{...}` are required for all messages,
+including ones that only consist of text.
+Delimiters of the resource format are required in addition to this,
+so messages may appear wrapped as e.g. `"{...}"`.
+
+Examples:
+
+```
+{Hello world}
+```
+
+```
+{Hello {$user}}
+```
+
+```
+input {$count :number minimumFractionDigits=1}
+{You have eaten {$count} apples}
+```
+
+```
+input {$count :number}
+match {$count}
+when 0 {You have no new message}
+when one {You have {$count} new message}
+when * {You have {$count} new messages}
+```
+
+```
+{ and some more}
+```
+
+### Start in text, encapsulate code
+
+The approach treats messages as template strings,
+which may include statements and expressions surrounded by `{...}`.
+Multi-variant messages require `match` and `when` statements that are followed by text at the top level.
+
+Whitespace around statements may need to be trimmed
+as e.g. `input` statements may be more readable when placed on a separate line,
+where they would be followed by a newline.
+At least the following trimming strategies may be considered:
+
+1. Do not trim any whitespace.
+1. Trim a minimal set of defined spaces:
+   - All spaces before and between variable statements.
+   - For single-variant messages, one newline after the last variable statement.
+   - For multivariant messages,
+     one space after a `when` statement and
+     one newline followed by any spaces before a subsequent `when` statement.
+1. Trim all leading and trailing whitespace.
+
+All "code" statements are surrounded by `{...}`,
+and all "text" is outside them.
+
+Simple messages are not surrounded by any delimiters
+other that what may be required by the resource format.
+
+Depending on the details of the syntax of code inside the `{...}`,
+unquoted non-numeric literals may need to be removed from the syntax.
+
+Examples using either "minimal" or "all" trimming:
+
+```
+Hello world
+```
+
+```
+Hello {$user}
+```
+
+```
+{input $count :number minimumFractionDigits=1}
+You have eaten {$count} apples
+```
+
+```
+{input $count :number}
+{match {$count}}
+{when 0} You have no new message
+{when one} You have {$count} new message
+{when *} You have {$count} new messages
+```
+
+```
+{| |}and some more
+```
```

Comment:
I will directly add some of what we did last night here via direct PR to the design doc.